File: 04-Java-Unit-Testing/Example-3/src/test/java/edu/odu/cs/cs350/examples/TestHtmlColor.java
package edu.odu.cs.cs350.examples;

import org.junit.jupiter.api.TestMethodOrder;
import org.junit.jupiter.api.MethodOrderer;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.BeforeEach;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

import static org.junit.jupiter.api.Assertions.*;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;
import org.hamcrest.core.IsNull;

/**
 * 1 - Does this piece of code perform the operations
 *     it was designed to perform?
 *
 * 2 - Does this piece of code do something it was not
 *     designed to perform?
 *
 * 1 Test per mutator
 */
@TestMethodOrder(MethodOrderer.MethodName.class)
public class TestHtmlColor {
    private HtmlColor black;
    private HtmlColor white;
    private HtmlColor red;
    private HtmlColor green;
    private HtmlColor blue;
    private HtmlColor rColor; // "random" color

    @BeforeEach
    public void setUp()
    {
        black  = new HtmlColor();
        white  = new HtmlColor(255, 255, 255);
        red    = new HtmlColor(255, 0, 0);
        green  = new HtmlColor(0, 255, 0);
        blue   = new HtmlColor(0, 0, 255);
        rColor = new HtmlColor(7, 62, 55);
    }

    @ParameterizedTest
    @ValueSource(ints = {-100, -5, -1, 0})
    public void testClampBelowRange(int value)
    {
        assertThat(HtmlColor.clamp(value), is(0));
    }

    @ParameterizedTest
    @ValueSource(ints = {0, 10, 100, 200, 250, 254, 255})
    public void testClampInRange(int value)
    {
        assertThat(HtmlColor.clamp(value), is(value));
    }

    @ParameterizedTest
    @ValueSource(ints = {256, 300, 700, 1000})
    public void testClampAboveRange(int value)
    {
        assertThat(HtmlColor.clamp(value), is(255));
    }

    @Test
    public void testDefaultConstructor()
    {
        HtmlColor color = new HtmlColor();

        assertThat(color.getRed(), is(0));
        assertThat(color.getBlue(), is(0));
        assertThat(color.getGreen(), is(0));

        assertThat(color.hashCode(), is(black.hashCode()));
        assertThat(color.toString(), equalTo("#000000"));
    }

    @Test
    public void testNonDefaultConstructor()
    {
        HtmlColor color = new HtmlColor(7, 62, 55);

        assertThat(color.getRed(), is(7));
        assertThat(color.getGreen(), is(62));
        assertThat(color.getBlue(), is(55));

        assertThat(rColor.hashCode(), is(color.hashCode()));

        String base16String = String.format(
            "#%02X%02X%02X",
            color.getRed(),
            color.getGreen(),
            color.getBlue()).toUpperCase();

        assertThat(color.toString(), equalTo(base16String));
    }

    @Test
    public void testNonDefaultConstructorInvalidRed()
    {
        HtmlColor color = new HtmlColor(-1, 62, 55);

        assertThat(color.getRed(), is(0));
        assertThat(color.getGreen(), is(62));
        assertThat(color.getBlue(), is(55));

        color = new HtmlColor(700, 62, 55);
        assertThat(color.getRed(), is(255));
        assertThat(color.getGreen(), is(62));
        assertThat(color.getBlue(), is(55));
    }

    @Test
    public void testNonDefaultConstructorInvalidGreen()
    {
        HtmlColor color = new HtmlColor(0, -1, 55);

        assertThat(color.getRed(), is(0));
        assertThat(color.getGreen(), is(0));
        assertThat(color.getBlue(), is(55));

        color = new HtmlColor(0, 620, 55);

        assertThat(color.getRed(), is(0));
        assertThat(color.getGreen(), is(255));
        assertThat(color.getBlue(), is(55));
    }

    @Test
    public void testNonDefaultConstructorInvalidBlue()
    {
        HtmlColor color = new HtmlColor(0, 0, -55);

        assertThat(color.getRed(), is(0));
        assertThat(color.getGreen(), is(0));
        assertThat(color.getBlue(), is(0));

        color = new HtmlColor(0, 62, 550);
        assertThat(color.getRed(), is(0));
        assertThat(color.getGreen(), is(62));
        assertThat(color.getBlue(), is(255));
    }

    @Test
    public void testSetRed()
    {
        HtmlColor color = (HtmlColor) black.clone();
        int oldHashCode = color.hashCode();

        color.setRed(100);

        assertThat(color.getRed(), is(100));
        assertThat(color.getBlue(), is(0));
        assertThat(color.getGreen(), is(0));

        assertThat(color, is(equalTo(color)));
        assertThat(color, is(not(equalTo(black))));
        assertThat(color.hashCode(), not(oldHashCode));
    }

    @Test
    public void testSetGreen()
    {
        HtmlColor color = (HtmlColor) blue.clone();
        int oldHashCode = color.hashCode();

        color.setGreen(100);

        assertThat(color.getRed(), is(0));
        assertThat(color.getBlue(), is(255));
        assertThat(color.getGreen(), is(100));

        assertThat(color, is(not(equalTo(blue))));
        assertThat(color.hashCode(), not(oldHashCode));
    }

    @Test
    public void testSetBlue()
    {
        HtmlColor color = (HtmlColor) white.clone();
        int oldHashCode = color.hashCode();

        color.setBlue(100);

        assertThat(color.getRed(), is(255));
        assertThat(color.getBlue(), is(100));
        assertThat(color.getGreen(), is(255));

        assertThat(color, is(not(equalTo(white))));
        assertThat(color.hashCode(), not(oldHashCode));
        assertThat(color.toString(), equalTo("#FFFF64"));

        //assertTrue(color.toString().contains("64"));
        assertThat(color.toString(), containsString("64"));
        assertThat(color.toString(), containsString("FFFF"));
    }
}

File: 04-Python-Unit-Testing/Example-2/tests/test_html_color.py
"""
1 - Does this piece of code perform the operations
    it was designed to perform?

2 - Does this piece of code do something it was not
    designed to perform?

1 Test per mutator
"""

import copy

import pytest
from hamcrest import *

from html_color import HtmlColor


@pytest.fixture
def black() -> HtmlColor:
    yield HtmlColor()


@pytest.fixture
def white() -> HtmlColor:
    yield HtmlColor(255, 255, 255)


@pytest.fixture
def blue() -> HtmlColor:
    yield HtmlColor(0, 0, 255)


@pytest.mark.parametrize("value", [-1, -100, -200, -250, -254, -255])
def test_clamp_below_range(value):
    assert_that(HtmlColor.clamp(value), is_(0))


@pytest.mark.parametrize("value", [0, 10, 100, 200, 250, 254, 255])
def test_clamp_in_range(value):
    assert_that(HtmlColor.clamp(value), is_(value))


@pytest.mark.parametrize("value", [256, 300, 700, 1000])
def test_clamp_above_range(value):
    assert_that(HtmlColor.clamp(value), is_(255))


def test_default_constructor(black: HtmlColor):
    color = HtmlColor()

    assert_that(color.red, is_(0))
    assert_that(color.green, is_(0))
    assert_that(color.blue, is_(0))

    assert_that(hash(color), is_(hash(black)))
    assert_that(str(color), equal_to("#000000"))


def test_non_default_constructor(black: HtmlColor):
    color = HtmlColor(7, 62, 55)

    assert_that(color.red, is_(7))
    assert_that(color.green, is_(62))
    assert_that(color.blue, is_(55))

    assert_that(hash(color), is_not(hash(black)))
    assert_that(str(color), equal_to("#073E37"))


@pytest.mark.parametrize("red_val", [256, 300, 1000])
def test_constructor_invalid_red(red_val):
    color = HtmlColor(red_val, 62, 55)

    assert_that(color.red, is_(255))
    assert_that(color.green, is_(62))
    assert_that(color.blue, is_(55))


def test_constructor_invalid_green():
    color = HtmlColor(1, -1, 55)

    assert_that(color.red, is_(1))
    assert_that(color.green, is_(0))
    assert_that(color.blue, is_(55))

    color = HtmlColor(1, 700, 55)

    assert_that(color.red, is_(1))
    assert_that(color.green, is_(255))
    assert_that(color.blue, is_(55))


def test_constructor_invalid_blue():
    color = HtmlColor(0, 0, -10)

    assert_that(color.red, is_(0))
    assert_that(color.green, is_(0))
    assert_that(color.blue, is_(0))

    color = HtmlColor(0, 0, 300)

    assert_that(color.red, is_(0))
    assert_that(color.green, is_(0))
    assert_that(color.blue, is_(255))


def test_set_red(black: HtmlColor):
    color = copy.deepcopy(black)
    old_hash_code = hash(color)

    color.red = 100

    assert_that(color.red, is_(100))
    assert_that(color.blue, is_(0))
    assert_that(color.green, is_(0))

    assert_that(color, is_(equal_to(color)))
    assert_that(color, is_not(equal_to(black)))
    assert_that(hash(color), is_not(old_hash_code))


def test_set_green(blue: HtmlColor):
    color = copy.deepcopy(blue)
    old_hash_code = hash(color)

    color.green = 100

    assert_that(color.red, is_(0))
    assert_that(color.green, is_(100))
    assert_that(color.blue, is_(255))

    assert_that(color, is_(equal_to(color)))
    assert_that(color, is_not(equal_to(blue)))
    assert_that(hash(color), is_not(old_hash_code))


def test_set_blue(white: HtmlColor):
    color = copy.deepcopy(white)
    old_hash_code = hash(color)

    color.blue = 100

    assert_that(color.red, is_(255))
    assert_that(color.green, is_(255))
    assert_that(color.blue, is_(100))

    assert_that(color, is_(equal_to(color)))
    assert_that(color, is_not(equal_to(white)))
    assert_that(hash(color), is_not(old_hash_code))

    assert_that(color, has_string("#FFFF64"))
